### PR TITLE
Remove CargoInternalsMetadataFetcher

### PR DESCRIPTION
**Pull request by smklein**
This seemed like a side-request of #147 ; CargoSubcommandMetadataFetcher is now the (only) default.

### CODE REVIEW

**Comment by kornholi** [comment]
One reason to keep using the old method is that it includes SHA256 hashes in the generated `crates.bzl`. They're not part of `cargo metadata` output but I'm not sure if that's an oversight or not.

#57

**Comment by acmcarther** [comment]
Thanks for putting this together.

Along with https://github.com/google/cargo-raze/pull/149#issuecomment-621493706, there's also a couple of issues in the cargo_metadata implementation that preclude deleting the cargo internals implementation immediately. 

https://github.com/google/cargo-raze/issues/136

Maybe also https://github.com/google/cargo-raze/issues/135

These urgently need attention (probably by me), sorry for being slow on this.

**Comment by smklein** [comment]
Question about that sha256 -- where is it being set for the package?

Digging around the cargo::core::summary implementation - which was previously a dependency here - the checksum is default "None", and is only set to anything else if explicitly set via "set_checksum".

Related: What's the best way to test this code? After the initial "let sha256" statement, I tried adding:

```
let sha256 = ...;

assert!(sha256.is_some());
assert!(sha256.is_none());
```

Just to see what blows up, and where. However, both `cargo test` and the smoke tests are passing. Is there another recommended way to get coverage here?

**Comment by smklein** [comment]
One of my pain points here was actually figuring out "how can I reproduce the checksum without the dependency on Cargo internals (or how can I move it into cargo_metadata to meet feature parity when using that dependency solution)?" In my naiveté, I figured that it might just be the sha256 of the Cargo.toml, but experimental evidence suggested otherwise. I'll document some of this process below; lemme know if it seems like I'm barking up the wrong tree.

With some minimal grepping in the cargo source tree, it looks like there are three spots where "set_checksum" is invoked, which actually would make it available for retrieval on the "package.manifest.summary()" object (of type cargo::core::summary::Summary).

1. cargo/sources/directory.rs: Within the impl of `Source` for `DirectorySource`, the `fn update(&mut self) -> CargoResult<()>` method looks like it iterates over all entries in the package root directory containing `Cargo.toml`. Seems like it looks for a file named `.cargo-checksum.json`, where it tries parsing out a checksum that was pre-calculated.

2. cargo/sources/registry/index.rs: In `IndexSummary`'s implementation of `fn parse(line &[u8], source_id: SourceId) -> CargoResult<IndexSummry>` (for parsing a single line from the registry index file of a package), the checksum looks it's being parsed from the Registry file format that is described here: https://doc.rust-lang.org/cargo/reference/registries.html#index-format

3. cargo/sources/registry/mod.rs: `RegistrySource` has a method, `fn get_pkg(&mut self, package: PackageId, path: &File) -> CargoResult<Package>`, which looks like it invokes `set_checksum` after extracting the checksum from a list of summaries. The checksums here look like they can originate from a "LocalRegistry" or a "RemoteRegistry".

- In "LocalRegistry", it looks like the checksum is calculated from a file named "NAME-VERSION.crate".

- In "RemoteRegistry", it looks like the checksum is calculated based on the data passed through the `data` argument to the `finish_download` method (seems like part of the `RegistryData` trait). I only see a meaningful value of `data` generated in `cargo/core/package.rs` (all other invocations are pass-through), where it looks like it's coming directly from the `Download`'s `data` field.

Through some experimental modifications to `cargo-raze/impl/src/metadata.rs`, it seems the sha256 value generated matches the sha256 of that "NAME-VERSION.crate" file which is generated by `cargo package`.

I see this file locally on my filesystem under `$HOME/.cargo/registry/cache/...`, but I'm not sure about what's the best way to access that information reliably, outside of the current invocation to `cargo::ops::resolve_ws_with_opts`.

**Comment by smklein** [comment]
The phrase "Please note that the internal structure of the Cargo home is not stabilized and may be subject to change at any time." within https://doc.rust-lang.org/cargo/guide/cargo-home.html makes me hesitant to use simple path substitution to access this crate. Odd that the value isn't propagated by `cargo metadata` natively though...

**Comment by smklein** [comment]
Hrm, maybe https://docs.rs/cargo-lock/4.0.1/cargo_lock/lockfile/index.html provides a solution here?

**Comment by smklein** [comment]
Okay, this is my current strategy:

- Parse metadata using cargo_metadata, as in #150 
- After that parsing is done -- which won't include the checksums, because `cargo metadata` doesn't actually parse lockfiles -- load the lockfile, and populate the SHA256 values for checksums if they're present when producing crate contexts.

Happy to dig into the other issues preventing us from having a single metadata fetcher! I'm pushing on this because (IMO) it makes cargo raze simpler and more stable, by having less dependencies on cargo's internal goop.

**Comment by smklein** [comment]
I believe with these changes, #57 should be done? @kornholi , does this approach work for you?

@acmcarther , do you have advice for testing the other issues?

**Comment by kornholi** [comment]
I also had the same approach in mind of parsing the lock file. Surprised how easy it was to plug in! :) I tested locally and it did produce the correct hashes.

Your analysis is correct! The hashes come from the packaged `.crate` files which are also what gets downloaded from crates.io. We want to pass them through to Bazel for reproducibility and caching reasons.

If Bazel knows what hash to expect for a remote URL, it can check that it matches what we get and also cache it without repeated fetches. The cache in Bazel is essentially a `content hash -> content` mapping, not `url -> content`.

I briefly looked at the `cargo metadata` code, and it doesn't look hard to expose checksums as part of the output. Deciding where that field would go is the hard part, and I assume it would need a version bump for their format. I'll open an issue.

**Comment by dfreese** [comment]
Is there anything else that's needed for this and #150 ?

They would be super helpful in terms of avoiding manually updating checksums.  :)

**Comment by acmcarther** [comment]
re: blockers on this and #150 

I'm looking into #136 and #135 now and will respond back here today after debugging those.

File: impl/src/planning.rs [hunk 1]
```diff
@@ -390,6 +410,9 @@ impl<'planner> WorkspaceSubplanner<'planner> {
         let own_crate_catalog_entry = self.crate_catalog.entry_for_package_id(&node.id).unwrap();
         let own_package = own_crate_catalog_entry.package();
 
+        let maybe_checksum =
```
[the remark this answers is not in the document]
Comment (smklein) [reply impl/src/planning.rs]: Done.

File: impl/src/planning.rs [hunk 2]
```diff
@@ -338,7 +340,7 @@ impl<'fetcher> BuildPlanner for BuildPlannerImpl<'fetcher> {
       platform_details: &platform_details,
     };
 
-    workspace_subplanner.produce_planned_build()
+    workspace_subplanner.produce_planned_build(&files)
```
[the remark this answers is not in the document]
Comment (smklein) [reply impl/src/planning.rs]: Updated to make a reference to files part of the WorkspacePlanner, and no longer passing it as an argument. I like this approach more!

Done.

**Comment by acmcarther** [comment]
I think this is ready to merge once synced to include https://github.com/google/cargo-raze/commit/73b87ab0ed49a3be35ec47cadd38c211e0c4470c (Thank you @smklein!), and [CI passes](https://travis-ci.org/github/google/cargo-raze/builds/686275069), which I expect will pass.

**Pull request merged**